contracts/randomNumber/RandomNoManager.sol

```diff
@@ -0,0 +1,88 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.0;
+import "../Core/interface/IParameters.sol";
+import "./IRandomNoClient.sol";
+import "./IRandomNoProvider.sol";
+import "../Initializable.sol";
+import "../lib/Random.sol";
+import "../Core/StateManager.sol";
+import "../Core/ACL.sol";
+import "./RandomNoStorage.sol";
+
+/**
+ *  @title : RandomNoManager
+ *  @notice : Allows clients to register for random no, and pull it once available
+ */
+
+contract RandomNoManager is Initializable, ACL, StateManager, RandomNoStorage, IRandomNoClient, IRandomNoProvider {
+    IParameters public parameters;
+
+    event RandomNumberAvailable(uint32 epoch);
+
+    /// @param blockManagerAddress The address of the BlockManager Contract
+    /// @param parametersAddress The address of the Parameters contract
+    function initialize(address blockManagerAddress, address parametersAddress) external initializer onlyRole(DEFAULT_ADMIN_ROLE) {
+        grantRole(SECRETS_MODIFIER_ROLE, blockManagerAddress);
+        parameters = IParameters(parametersAddress);
+    }
+
+    /// @notice Allows Client to register for random number
+    /// Per request a rquest id is generated, which is binded to one epoch
+    /// this epoch is current epoch if Protocol is in commit state, and epoch + 1 if in any other states
+    /// @return requestId : unique request id
+    function register() external override initialized returns (bytes32 requestId) {
+        uint32 epoch = getEpoch(parameters.epochLength());
+        State state = getState(parameters.epochLength());
```

Comment:
two repeated external calls. Store it in memory instead

Reply:
done